src/Features/Core/Portable/TodoComments/AbstractTodoCommentService.cs

```diff
@@ -8,11 +8,24 @@
 using Microsoft.CodeAnalysis.Remote;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.TodoComments
 {
     internal abstract class AbstractTodoCommentService : ITodoCommentService
     {
+        // we hold onto workspace to make sure given input (Document) belong to right workspace.
+        // since remote host is from workspace service, different workspace can have different expectation
+        // on remote host, so we need to make sure given input always belong to right workspace where
+        // the session belong to.
+        private readonly Workspace _workspace;
+        private KeepAliveSessionHolder _session;
```

Comment:
so that it doesnt need to create new connection each time. reusing connection is about 10-100 times faster it looks like. once I finish all changes so that I can pool session, it could go back to regular one. or stay as keep alive.